src/filesystem/NetworkXioServer.cpp

```diff
@@ -0,0 +1,587 @@
+}
+
+NetworkXioClientData*
+NetworkXioServer::allocate_client_data()
+{
+    try
+    {
+        NetworkXioClientData *cd = new NetworkXioClientData;
+        cd->disconnected = false;
+        cd->refcnt = 0;
+        cd->mpool = xio_mpool;
+        cd->server = this;
+        return cd;
+    }
+    catch (const std::bad_alloc&)
+    {
+        return NULL;
+    }
+}
+
+int
+NetworkXioServer::create_session_connection(xio_session *session,
+                                            xio_session_event_data *evdata)
+{
+    NetworkXioClientData *cd = allocate_client_data();
+    if (cd)
+    {
+        try
+        {
+            NetworkXioIOHandler *ioh_ptr = new NetworkXioIOHandler(fs_,
+                                                                   wq_);
+            cd->ioh = ioh_ptr;
+            cd->session = session;
+            cd->conn = evdata->conn;
+        }
+        catch (...)
+        {
+            LOG_ERROR("cannot create IO handler");
+            delete cd;
+            return -1;
+        }
+        xio_connection_attr xconattr;
+        xconattr.user_context = cd;
+        (void) xio_modify_connection(evdata->conn,
+                                     &xconattr,
+                                     XIO_CONNECTION_ATTR_USER_CTX);
+        wq_->open_sessions_inc();
+        return 0;
+    }
+    LOG_ERROR("cannot allocate client data");
+    return -1;
+}
+
+void
+NetworkXioServer::destroy_session_connection(xio_session *session ATTR_UNUSED,
+                                             xio_session_event_data *evdata)
+{
+    auto cd = static_cast<NetworkXioClientData*>(evdata->conn_user_context);
+    cd->disconnected = true;
+    if (!cd->refcnt)
+    {
+        xio_connection_destroy(cd->conn);
+        wq_->open_sessions_dec();
+        delete cd->ioh;
+        delete cd;
+    }
+}
+
+int
+NetworkXioServer::on_new_session(xio_session *session,
+                                 xio_new_session_req * /*req*/)
+{
+    if (xio_accept(session, NULL, 0, NULL, 0) < 0)
+    {
+        LOG_ERROR("cannot accept new session, error: "
+                  << xio_strerror(xio_errno()));
+    }
+    return 0;
+}
+
+int
+NetworkXioServer::on_session_event(xio_session *session,
+                                   xio_session_event_data *event_data)
+{
+    switch (event_data->event)
+    {
+    case XIO_SESSION_NEW_CONNECTION_EVENT:
+        create_session_connection(session, event_data);
+        break;
+    case XIO_SESSION_CONNECTION_TEARDOWN_EVENT:
+        destroy_session_connection(session, event_data);
+        break;
+    case XIO_SESSION_TEARDOWN_EVENT:
+        xio_session_destroy(session);
+        break;
+    default:
+        break;
+    };
+    return 0;
+}
+
+int
+NetworkXioServer::assign_data_in_buf(struct xio_msg *msg)
+{
+    xio_iovec_ex *isglist = vmsg_sglist(&msg->in);
+
+    LOG_DEBUG("assign buffer, len: " << isglist[0].iov_len);
+
+    xio_reg_mem in_xbuf;
+    xio_mem_alloc(isglist[0].iov_len, &in_xbuf);
```

Comment:
xio_mem_alloc could fail.
Since it allocates and registers memory and we're in the data path, is that not going to be too expensive?